### PR TITLE
Created Dijkstra's Two Stack Algorithm

### PR DESCRIPTION
**Pull request by echoaj**
### **Describe your change:**

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### **Checklist:**
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [x] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: data_structures/stacks/dijkstras_two_stack_algorithm.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+NOTE:   It only works with positive numbers.
+"""
+__author__ = "Alexander Joslin"
+
+
+class Stack:
+    """ Stack Data Structure Used for the Algorithm"""
+    def __init__(self):
+        self.__array = []
+        self.__len = 0
```
Comment (cclauss) [inline data_structures/stacks/dijkstras_two_stack_algorithm.py:39]: Let’s do as [`__len__()`](https://docs.python.org/3/reference/datamodel.html#object.__len__)

File: data_structures/stacks/dijkstras_two_stack_algorithm.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+        self.underflow_error = False
+
+    def is_empty(self):
```
Comment (cclauss) [inline data_structures/stacks/dijkstras_two_stack_algorithm.py:42]: We don’t need this method.  `if stack:` is sufficient.

File: data_structures/stacks/dijkstras_two_stack_algorithm.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+        if self.__len == 0:
+            return True
+        else:
+            return False
+
+    def push(self, x):
+        self.__array.append(x)
+        self.__len += 1
```
Comment (cclauss) [inline data_structures/stacks/dijkstras_two_stack_algorithm.py:50]: We don’t to need to do this busywork if `__len__()` returns `len(self.array)`.

File: data_structures/stacks/dijkstras_two_stack_algorithm.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+
+    def pop(self):
+        if not self.is_empty():
```
Comment (cclauss) [inline data_structures/stacks/dijkstras_two_stack_algorithm.py:53]:
```suggestion
        if not self.array:
            raise IndexError("pop from empty list")
```
just like [].pop()

File: data_structures/stacks/dijkstras_two_stack_algorithm.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+            self.__array.pop()
+            self.__len -= 1
+            self.underflow_error = False
+        else:
+            self.underflow_error = True
+
+    def peek(self):
+        return self.__array[-1] if not self.is_empty() else None
+
+    def length(self):
```
Comment (cclauss) [inline data_structures/stacks/dijkstras_two_stack_algorithm.py:63]: Not in Python!

File: data_structures/stacks/dijkstras_two_stack_algorithm.py [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+        return self.__len
+
+    def print_stack(self):
+        print(self.__array)
```
Comment (cclauss) [inline data_structures/stacks/dijkstras_two_stack_algorithm.py:67]: Define as [`__str__()`](https://docs.python.org/3/reference/datamodel.html#object.__str__)

File: data_structures/stacks/dijkstras_two_stack_algorithm.py [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+            # RULE 4
+            opr1 = operator_stack.peek()
+            operator_stack.pop()
+            num1 = operand_stack.peek()
+            operand_stack.pop()
+            num2 = operand_stack.peek()
+            operand_stack.pop()
+
+            total = 0
+            if opr1 == "+":
```
Comment (cclauss) [inline data_structures/stacks/dijkstras_two_stack_algorithm.py:103]:
```python
import operator as op
    operators = {
        "*": op.mul,
        "/": op.true_div,
        "+": op.add,
        "-": op.sub,
    }

File: data_structures/stacks/dijkstras_two_stack_algorithm.py [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+            elif opr1 == "%":
+                total = num2 % num1
+
+            operand_stack.push(total)
+        else:
+            # RULE 3
+            pass
+
+    # RULE 5
+    result = operand_stack.peek()
```
Comment (cclauss) [inline data_structures/stacks/dijkstras_two_stack_algorithm.py:120]:
```suggestion
    return operand_stack.peek()
```

File: data_structures/stacks/dijkstras_two_stack_algorithm.py [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+    return result
+
+
+def main():
+    equation = "(5 + ((4 * 2) * (2 + 3)))"
+    # answer = 45
+    answer = dijkstras_two_stack_algorithm(equation)
```
Comment (cclauss) [inline data_structures/stacks/dijkstras_two_stack_algorithm.py:127]:
```suggestion
    print(f"{dijkstras_two_stack_algorithm(equation) = }")
```

File: data_structures/stacks/dijkstras_two_stack_algorithm.py [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+        return self.__array[-1] if not self.is_empty() else None
+
+    def length(self):
+        return self.__len
+
+    def print_stack(self):
+        print(self.__array)
+
+
+def dijkstras_two_stack_algorithm(equation):
```
Comment (cclauss) [inline data_structures/stacks/dijkstras_two_stack_algorithm.py:70]: Type hints?

File: data_structures/stacks/dijkstras_two_stack_algorithm.py (hunk 7 above)
Comment (echoaj) [reply data_structures/stacks/dijkstras_two_stack_algorithm.py:103]: I don't think this is necessary.  According to the contributing guidelines importing external libraries should be avoided.
"Avoid importing external libraries for basic algorithms. Only use those libraries for complicated algorithms."

Comment (cclauss) [reply data_structures/stacks/dijkstras_two_stack_algorithm.py:70]: The Python standard library is not an _external_ library.

File: data_structures/stacks/dijkstras_two_stack_algorithm.py [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+        for use as an operand of the next operator popped off the operator stack.
+
+RULE 5: When the entire infix expression has been scanned, the value left on
+        the operand stack represents the value of the expression.
+
+NOTE:   It only works with whole numbers.
+"""
+__author__ = "Alexander Joslin"
+
+from stack import Stack
```
Comment (cclauss) [inline data_structures/stacks/dijkstras_two_stack_algorithm.py:34]:
```suggestion
from .stack import Stack
```

**Review by cclauss — APPROVED** [review]
Nice work!

**Pull request merged**